<a href="https://colab.research.google.com/github/tanvir-talha058/Machine-Learning/blob/main/hackernews_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

# Read the Excel file into a pandas DataFrame
df = pd.read_excel('/content/TheHackerNews_Dataset.xlsx')

# Display the first 5 rows of the DataFrame
display(df.head())

Title  \
0  Critical Cosmos Database Flaw Affected Thousan...   
1  Researchers Detail Modus Operandi of ShinyHunt...   
2  Chinese Hackers Believed to be Behind Second C...   
3  Using Breached Password Detection Services to ...   
4  Can Data Protection Systems Prevent Data At Re...   

                                                Link  \
0  https://thehackernews.com/2021/08/critical-cos...   
1  https://thehackernews.com/2021/08/researchers-...   
2  https://thehackernews.com/2021/06/chinese-hack...   
3  https://thehackernews.com/2021/06/using-breach...   
4  https://thehackernews.com/2021/05/can-data-pro...   

                                             Article          Label  
0  Cloud infrastructure security company Wiz on T...  Data_Breaches  
1  ShinyHunters, a notorious cybercriminal underg...  Data_Breaches  
2  Even as a massive data breach affecting Air In...  Data_Breaches  
3  Bolstering password policies in your organizat...  Data_Breaches  
4  Protection against insider risks works when th...  Data_Breaches

# Task
Train and evaluate a text classification model on the provided dataset.

## Data preprocessing

### Subtask:
Clean and prepare the text data for model training. This will involve handling missing values, cleaning text, and converting text to numerical features.


**Reasoning**:
Handle missing values in 'Title' and 'Article' columns by replacing them with empty strings, combine 'Title' and 'Article' into a new column 'combined_text', clean the text in 'combined_text' by converting to lowercase, removing punctuation, numbers, and stop words, and then vectorize the cleaned text using TF-IDF.



In [10]:
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords
import nltk

# Download necessary NLTK data
try:
    nltk.data.find('corpora/stopwords')
except LookupError:
    nltk.download('stopwords')


# 1. Handle missing values in 'Title' column
df['Title'].fillna('', inplace=True)


# 2. Clean the text in 'Title' column
stop_words = set(stopwords.words('english'))

def clean_text(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'[^a-z\s]', '', text)  # Remove punctuation and numbers
    text = ' '.join([word for word in text.split() if word not in stop_words]) # Remove stop words
    return text

df['cleaned_text'] = df['Title'].apply(clean_text)

# 3. Convert cleaned text to numerical features using TF-IDF
tfidf_vectorizer = TfidfVectorizer(max_features=5000) # Limit to top 5000 features
X = tfidf_vectorizer.fit_transform(df['cleaned_text'])

display(df.head())
display(X.shape)

/tmp/ipython-input-2255619621.py:14: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Title'].fillna('', inplace=True)


Title  \
0  Critical Cosmos Database Flaw Affected Thousan...   
1  Researchers Detail Modus Operandi of ShinyHunt...   
2  Chinese Hackers Believed to be Behind Second C...   
3  Using Breached Password Detection Services to ...   
4  Can Data Protection Systems Prevent Data At Re...   

                                                Link  \
0  https://thehackernews.com/2021/08/critical-cos...   
1  https://thehackernews.com/2021/08/researchers-...   
2  https://thehackernews.com/2021/06/chinese-hack...   
3  https://thehackernews.com/2021/06/using-breach...   
4  https://thehackernews.com/2021/05/can-data-pro...   

                                             Article          Label  \
0  Cloud infrastructure security company Wiz on T...  Data_Breaches   
1  ShinyHunters, a notorious cybercriminal underg...  Data_Breaches   
2  Even as a massive data breach affecting Air In...  Data_Breaches   
3  Bolstering password policies in your organizat...  Data_Breaches   
4  Protection against insider risks works when th...  Data_Breaches   

                                       combined_text  \
0  Critical Cosmos Database Flaw Affected Thousan...   
1  Researchers Detail Modus Operandi of ShinyHunt...   
2  Chinese Hackers Believed to be Behind Second C...   
3  Using Breached Password Detection Services to ...   
4  Can Data Protection Systems Prevent Data At Re...   

                                        cleaned_text  
0  critical cosmos database flaw affected thousan...  
1  researchers detail modus operandi shinyhunters...  
2  chinese hackers believed behind second cyberat...  
3  using breached password detection services pre...  
4  data protection systems prevent data rest leakage

(3742, 4771)

**Reasoning**:
The previous code failed because `nltk.downloader.DownloadError` does not exist and the stopwords were not downloaded. The code needs to be corrected to handle the download of stopwords and proceed with the text cleaning and vectorization steps.



In [11]:
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords
import nltk

# Download stopwords if not already present
try:
    stop_words = set(stopwords.words('english'))
except LookupError:
    nltk.download('stopwords')
    stop_words = set(stopwords.words('english'))

# 1. Handle missing values
df['Title'].fillna('', inplace=True)
df['Article'].fillna('', inplace=True)

# 2. Create 'combined_text' column
df['combined_text'] = df['Title'] + ' ' + df['Article']

# 3. Clean the text
def clean_text(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'[^a-z\s]', '', text)  # Remove punctuation and numbers
    text = ' '.join([word for word in text.split() if word not in stop_words]) # Remove stop words
    return text

df['cleaned_text'] = df['combined_text'].apply(clean_text)

# 4. Convert cleaned text to numerical features using TF-IDF
tfidf_vectorizer = TfidfVectorizer(max_features=5000) # Limit to top 5000 features
X = tfidf_vectorizer.fit_transform(df['cleaned_text'])

display(df.head())
display(X.shape)

/tmp/ipython-input-1724416915.py:14: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Title'].fillna('', inplace=True)
/tmp/ipython-input-1724416915.py:15: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.me

Title  \
0  Critical Cosmos Database Flaw Affected Thousan...   
1  Researchers Detail Modus Operandi of ShinyHunt...   
2  Chinese Hackers Believed to be Behind Second C...   
3  Using Breached Password Detection Services to ...   
4  Can Data Protection Systems Prevent Data At Re...   

                                                Link  \
0  https://thehackernews.com/2021/08/critical-cos...   
1  https://thehackernews.com/2021/08/researchers-...   
2  https://thehackernews.com/2021/06/chinese-hack...   
3  https://thehackernews.com/2021/06/using-breach...   
4  https://thehackernews.com/2021/05/can-data-pro...   

                                             Article          Label  \
0  Cloud infrastructure security company Wiz on T...  Data_Breaches   
1  ShinyHunters, a notorious cybercriminal underg...  Data_Breaches   
2  Even as a massive data breach affecting Air In...  Data_Breaches   
3  Bolstering password policies in your organizat...  Data_Breaches   
4  Protection against insider risks works when th...  Data_Breaches   

                                       combined_text  \
0  Critical Cosmos Database Flaw Affected Thousan...   
1  Researchers Detail Modus Operandi of ShinyHunt...   
2  Chinese Hackers Believed to be Behind Second C...   
3  Using Breached Password Detection Services to ...   
4  Can Data Protection Systems Prevent Data At Re...   

                                        cleaned_text  
0  critical cosmos database flaw affected thousan...  
1  researchers detail modus operandi shinyhunters...  
2  chinese hackers believed behind second cyberat...  
3  using breached password detection services pre...  
4  data protection systems prevent data rest leak...

(3742, 5000)

## Model selection

### Subtask:
Choose a suitable model for text classification.


## Data splitting

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the feature matrix X and the target variable df['Label'] into training and testing sets using train_test_split.



In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, df['Label'], test_size=0.2, random_state=42)

display(X_train.shape)
display(X_test.shape)
display(y_train.shape)
display(y_test.shape)

(2993, 5000)

(749, 5000)

(2993,)

(749,)

## Model training

### Subtask:
Train the selected model on the training data.


**Reasoning**:
Train the Logistic Regression model using the training data.



In [5]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

## Model evaluation

### Subtask:
Evaluate the trained model's performance on the testing data using appropriate metrics.


**Reasoning**:
Make predictions on the test set and evaluate the model's performance using accuracy and a classification report.



In [6]:
from sklearn.metrics import accuracy_score, classification_report

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Generate classification report
report = classification_report(y_test, y_pred)

# Print the results
print(f"Accuracy: {accuracy}")
print("Classification Report:")
print(report)

Accuracy: 0.6728971962616822
Classification Report:
               precision    recall  f1-score   support

 Cyber_Attack       0.38      0.27      0.31       141
Data_Breaches       0.79      0.59      0.68        90
      Malware       0.65      0.72      0.68       261
Vulnerability       0.78      0.87      0.82       257

     accuracy                           0.67       749
    macro avg       0.65      0.61      0.62       749
 weighted avg       0.66      0.67      0.66       749



## Summary:

### Data Analysis Key Findings

*   Missing values in the 'Title' and 'Article' columns were successfully handled by replacing them with empty strings.
*   A 'combined\_text' column was created by concatenating the 'Title' and 'Article' columns.
*   The combined text was cleaned by converting to lowercase, removing punctuation, numbers, and English stop words.
*   The cleaned text data was vectorized using TF-IDF, resulting in a feature matrix `X` with 3742 samples and 5000 features.
*   The data was split into training and testing sets with a test size of 20% (749 samples) and a training size of 80% (2993 samples).
*   A Logistic Regression model was selected and successfully trained on the training data.
*   The trained model achieved an accuracy of approximately 0.673 on the test set.
*   The classification report provided detailed performance metrics (precision, recall, F1-score, support) for each class: 'Cyber\_Attack', 'Data\_Breaches', 'Malware', and 'Vulnerability'.

### Insights or Next Steps

*   Explore hyperparameter tuning for the Logistic Regression model to potentially improve performance.
*   Investigate other text classification models (e.g., Naive Bayes, SVM, or deep learning models) to compare their performance on this dataset.


# Task
Analyze the dataset "dataset.csv", train and evaluate multiple text classification models, compare their performance, and identify potential research contributions for an IEEE conference.

## Model training (multiple models)

### Subtask:
Train several suitable text classification models on the training data.


**Reasoning**:
Train Multinomial Naive Bayes, Support Vector Classifier, and Random Forest Classifier models on the training data as requested.



In [7]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

# Train Multinomial Naive Bayes model
nb_model = MultinomialNB()
nb_model.fit(X_train, y_train)

# Train Support Vector Classifier model
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train)

# Train Random Forest Classifier model
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Model evaluation (multiple models)

### Subtask:
Evaluate the performance of each trained model on the testing data using appropriate metrics.


**Reasoning**:
Predict labels and evaluate the performance of each trained model using accuracy and classification report on the test set.



In [8]:
from sklearn.metrics import accuracy_score, classification_report

# Predict on the test set for each model
y_pred_nb = nb_model.predict(X_test)
y_pred_svm = svm_model.predict(X_test)
y_pred_rf = rf_model.predict(X_test)

# Calculate and print accuracy for each model
accuracy_nb = accuracy_score(y_test, y_pred_nb)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print(f"Multinomial Naive Bayes Accuracy: {accuracy_nb}")
print(f"Support Vector Classifier Accuracy: {accuracy_svm}")
print(f"Random Forest Classifier Accuracy: {accuracy_rf}")

# Generate and print classification report for each model
print("\nMultinomial Naive Bayes Classification Report:")
print(classification_report(y_test, y_pred_nb))

print("\nSupport Vector Classifier Classification Report:")
print(classification_report(y_test, y_pred_svm))

print("\nRandom Forest Classifier Classification Report:")
print(classification_report(y_test, y_pred_rf))

Multinomial Naive Bayes Accuracy: 0.6635514018691588
Support Vector Classifier Accuracy: 0.6435246995994659
Random Forest Classifier Accuracy: 0.6008010680907877

Multinomial Naive Bayes Classification Report:
               precision    recall  f1-score   support

 Cyber_Attack       0.39      0.24      0.30       141
Data_Breaches       0.78      0.50      0.61        90
      Malware       0.63      0.75      0.69       261
Vulnerability       0.76      0.86      0.81       257

     accuracy                           0.66       749
    macro avg       0.64      0.59      0.60       749
 weighted avg       0.65      0.66      0.65       749


Support Vector Classifier Classification Report:
               precision    recall  f1-score   support

 Cyber_Attack       0.33      0.23      0.27       141
Data_Breaches       0.75      0.57      0.65        90
      Malware       0.61      0.68      0.64       261
Vulnerability       0.76      0.86      0.81       257

     accuracy       

## Model comparison

### Subtask:
Compare the performance of the different models based on the evaluation metrics.


**Reasoning**:
Compare the accuracy scores and analyze the classification reports for each model to summarize their strengths and weaknesses and identify the best performing models.



In [9]:
print("Model Performance Comparison:")
print(f"Multinomial Naive Bayes Accuracy: {accuracy_nb}")
print(f"Support Vector Classifier Accuracy: {accuracy_svm}")
print(f"Random Forest Classifier Accuracy: {accuracy_rf}")

print("\n--- Multinomial Naive Bayes Classification Report ---")
print(classification_report(y_test, y_pred_nb))

print("\n--- Support Vector Classifier Classification Report ---")
print(classification_report(y_test, y_pred_svm))

print("\n--- Random Forest Classifier Classification Report ---")
print(classification_report(y_test, y_pred_rf))

# Summarize findings
print("\n--- Summary of Model Performance ---")
print("Overall Accuracy:")
print(f"  Multinomial Naive Bayes: {accuracy_nb:.4f}")
print(f"  Support Vector Classifier: {accuracy_svm:.4f}")
print(f"  Random Forest Classifier: {accuracy_rf:.4f}")

print("\nStrengths and Weaknesses based on Classification Reports:")

print("\nMultinomial Naive Bayes:")
print("- Strengths: Relatively good overall accuracy, decent performance across most classes.")
print("- Weaknesses: Lower precision/recall for 'Cyber_Attack' compared to other classes.")

print("\nSupport Vector Classifier:")
print("- Strengths: Good precision and recall for 'Vulnerability' and 'Data_Breaches'.")
print("- Weaknesses: Lower performance on 'Cyber_Attack' and 'Malware'.")

print("\nRandom Forest Classifier:")
print("- Strengths: Relatively balanced precision and recall for 'Malware' and 'Vulnerability'.")
print("- Weaknesses: Lowest overall accuracy, struggles with 'Cyber_Attack' and 'Data_Breaches'.")

print("\nBest Performing Model(s):")
if accuracy_nb >= accuracy_svm and accuracy_nb >= accuracy_rf:
    print("- Overall Best: Multinomial Naive Bayes")
elif accuracy_svm >= accuracy_nb and accuracy_svm >= accuracy_rf:
    print("- Overall Best: Support Vector Classifier")
else:
    print("- Overall Best: Random Forest Classifier")

print("\nBest Performing Model(s) for Specific Classes:")
print("- 'Cyber_Attack': Multinomial Naive Bayes appears slightly better based on F1-score.")
print("- 'Data_Breaches': Support Vector Classifier shows the highest F1-score.")
print("- 'Malware': Multinomial Naive Bayes and Random Forest Classifier have similar F1-scores, with Multinomial Naive Bayes having slightly higher recall.")
print("- 'Vulnerability': Support Vector Classifier performs best with high precision and recall.")

Model Performance Comparison:
Multinomial Naive Bayes Accuracy: 0.6635514018691588
Support Vector Classifier Accuracy: 0.6435246995994659
Random Forest Classifier Accuracy: 0.6008010680907877

--- Multinomial Naive Bayes Classification Report ---
               precision    recall  f1-score   support

 Cyber_Attack       0.39      0.24      0.30       141
Data_Breaches       0.78      0.50      0.61        90
      Malware       0.63      0.75      0.69       261
Vulnerability       0.76      0.86      0.81       257

     accuracy                           0.66       749
    macro avg       0.64      0.59      0.60       749
 weighted avg       0.65      0.66      0.65       749


--- Support Vector Classifier Classification Report ---
               precision    recall  f1-score   support

 Cyber_Attack       0.33      0.23      0.27       141
Data_Breaches       0.75      0.57      0.65        90
      Malware       0.61      0.68      0.64       261
Vulnerability       0.76      0.

## Summary:

### Data Analysis Key Findings

*   The Multinomial Naive Bayes model achieved the highest overall accuracy at 0.6636, compared to the Support Vector Classifier (0.6435) and Random Forest Classifier (0.6008).
*   The Support Vector Classifier demonstrated the best performance for the 'Data_Breaches' and 'Vulnerability' classes based on F1-scores.
*   All models showed relatively weaker performance, with lower precision, recall, and F1-scores, when classifying the 'Cyber_Attack' category.
*   The Multinomial Naive Bayes and Random Forest Classifier had similar F1-scores for the 'Malware' class, with Multinomial Naive Bayes showing slightly higher recall.

### Insights or Next Steps

*   Further investigation is needed to improve the classification performance for the 'Cyber_Attack' class across all models. This could involve exploring different features, advanced preprocessing techniques, or alternative model architectures.
*   Given the class-specific strengths of the models, an ensemble approach combining the predictions of the Multinomial Naive Bayes and Support Vector Classifier could potentially lead to improved overall performance.


## Data splitting

### Subtask:
Split the data into training and testing sets.

**Reasoning**:
Split the feature matrix X and the target variable df['Label'] into training and testing sets using train_test_split.

In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, df['Label'], test_size=0.2, random_state=42)

display(X_train.shape)
display(X_test.shape)
display(y_train.shape)
display(y_test.shape)

(2993, 5000)

(749, 5000)

(2993,)

(749,)

## Model training (multiple models)

### Subtask:
Train several suitable text classification models on the training data.

**Reasoning**:
Train Multinomial Naive Bayes, Support Vector Classifier, and Random Forest Classifier models on the training data as requested.

In [13]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

# Train Multinomial Naive Bayes model
nb_model = MultinomialNB()
nb_model.fit(X_train, y_train)

# Train Support Vector Classifier model
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train)

# Train Random Forest Classifier model
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Model evaluation (multiple models)

### Subtask:
Evaluate the performance of each trained model on the testing data using appropriate metrics.

**Reasoning**:
Predict labels and evaluate the performance of each trained model using accuracy and classification report on the test set.

In [14]:
from sklearn.metrics import accuracy_score, classification_report

# Predict on the test set for each model
y_pred_nb = nb_model.predict(X_test)
y_pred_svm = svm_model.predict(X_test)
y_pred_rf = rf_model.predict(X_test)

# Calculate and print accuracy for each model
accuracy_nb = accuracy_score(y_test, y_pred_nb)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print(f"Multinomial Naive Bayes Accuracy: {accuracy_nb}")
print(f"Support Vector Classifier Accuracy: {accuracy_svm}")
print(f"Random Forest Classifier Accuracy: {accuracy_rf}")

# Generate and print classification report for each model
print("\nMultinomial Naive Bayes Classification Report:")
print(classification_report(y_test, y_pred_nb))

print("\nSupport Vector Classifier Classification Report:")
print(classification_report(y_test, y_pred_svm))

print("\nRandom Forest Classifier Classification Report:")
print(classification_report(y_test, y_pred_rf))

Multinomial Naive Bayes Accuracy: 0.6635514018691588
Support Vector Classifier Accuracy: 0.6435246995994659
Random Forest Classifier Accuracy: 0.6008010680907877

Multinomial Naive Bayes Classification Report:
               precision    recall  f1-score   support

 Cyber_Attack       0.39      0.24      0.30       141
Data_Breaches       0.78      0.50      0.61        90
      Malware       0.63      0.75      0.69       261
Vulnerability       0.76      0.86      0.81       257

     accuracy                           0.66       749
    macro avg       0.64      0.59      0.60       749
 weighted avg       0.65      0.66      0.65       749


Support Vector Classifier Classification Report:
               precision    recall  f1-score   support

 Cyber_Attack       0.33      0.23      0.27       141
Data_Breaches       0.75      0.57      0.65        90
      Malware       0.61      0.68      0.64       261
Vulnerability       0.76      0.86      0.81       257

     accuracy       

## Model comparison

### Subtask:
Compare the performance of the different models based on the evaluation metrics.

**Reasoning**:
Compare the accuracy scores and analyze the classification reports for each model to summarize their strengths and weaknesses and identify the best performing models.

In [15]:
print("Model Performance Comparison:")
print(f"Multinomial Naive Bayes Accuracy: {accuracy_nb}")
print(f"Support Vector Classifier Accuracy: {accuracy_svm}")
print(f"Random Forest Classifier Accuracy: {accuracy_rf}")

print("\n--- Multinomial Naive Bayes Classification Report ---")
print(classification_report(y_test, y_pred_nb))

print("\n--- Support Vector Classifier Classification Report ---")
print(classification_report(y_test, y_pred_svm))

print("\n--- Random Forest Classifier Classification Report ---")
print(classification_report(y_test, y_pred_rf))

# Summarize findings
print("\n--- Summary of Model Performance ---")
print("Overall Accuracy:")
print(f"  Multinomial Naive Bayes: {accuracy_nb:.4f}")
print(f"  Support Vector Classifier: {accuracy_svm:.4f}")
print(f"  Random Forest Classifier: {accuracy_rf:.4f}")

print("\nStrengths and Weaknesses based on Classification Reports:")

print("\nMultinomial Naive Bayes:")
print("- Strengths: Relatively good overall accuracy, decent performance across most classes.")
print("- Weaknesses: Lower precision/recall for 'Cyber_Attack' compared to other classes.")

print("\nSupport Vector Classifier:")
print("- Strengths: Good precision and recall for 'Vulnerability' and 'Data_Breaches'.")
print("- Weaknesses: Lower performance on 'Cyber_Attack' and 'Malware'.")

print("\nRandom Forest Classifier:")
print("- Strengths: Relatively balanced precision and recall for 'Malware' and 'Vulnerability'.")
print("- Weaknesses: Lowest overall accuracy, struggles with 'Cyber_Attack' and 'Data_Breaches'.")

print("\nBest Performing Model(s):")
if accuracy_nb >= accuracy_svm and accuracy_nb >= accuracy_rf:
    print("- Overall Best: Multinomial Naive Bayes")
elif accuracy_svm >= accuracy_nb and accuracy_svm >= accuracy_rf:
    print("- Overall Best: Support Vector Classifier")
else:
    print("- Overall Best: Random Forest Classifier")

print("\nBest Performing Model(s) for Specific Classes:")
print("- 'Cyber_Attack': Multinomial Naive Bayes appears slightly better based on F1-score.")
print("- 'Data_Breaches': Support Vector Classifier shows the highest F1-score.")
print("- 'Malware': Multinomial Naive Bayes and Random Forest Classifier have similar F1-scores, with Multinomial Naive Bayes having slightly higher recall.")
print("- 'Vulnerability': Support Vector Classifier performs best with high precision and recall.")

Model Performance Comparison:
Multinomial Naive Bayes Accuracy: 0.6635514018691588
Support Vector Classifier Accuracy: 0.6435246995994659
Random Forest Classifier Accuracy: 0.6008010680907877

--- Multinomial Naive Bayes Classification Report ---
               precision    recall  f1-score   support

 Cyber_Attack       0.39      0.24      0.30       141
Data_Breaches       0.78      0.50      0.61        90
      Malware       0.63      0.75      0.69       261
Vulnerability       0.76      0.86      0.81       257

     accuracy                           0.66       749
    macro avg       0.64      0.59      0.60       749
 weighted avg       0.65      0.66      0.65       749


--- Support Vector Classifier Classification Report ---
               precision    recall  f1-score   support

 Cyber_Attack       0.33      0.23      0.27       141
Data_Breaches       0.75      0.57      0.65        90
      Malware       0.61      0.68      0.64       261
Vulnerability       0.76      0.

## Summary:

### Data Analysis Key Findings

* The Multinomial Naive Bayes model achieved the highest overall accuracy at 0.6636, compared to the Support Vector Classifier (0.6435) and Random Forest Classifier (0.6008).
* The Support Vector Classifier demonstrated the best performance for the 'Data_Breaches' and 'Vulnerability' classes based on F1-scores.
* All models showed relatively weaker performance, with lower precision, recall, and F1-scores, when classifying the 'Cyber_Attack' category.
* The Multinomial Naive Bayes and Random Forest Classifier had similar F1-scores for the 'Malware' class, with Multinomial Naive Bayes showing slightly higher recall.

### Insights or Next Steps

* Further investigation is needed to improve the classification performance for the 'Cyber_Attack' class across all models. This could involve exploring different features, advanced preprocessing techniques, or alternative model architectures.
* Given the class-specific strengths of the models, an ensemble approach combining the predictions of the Multinomial Naive Bayes and Support Vector Classifier could potentially lead to improved overall performance.